### PR TITLE
Feature/newborn guardians

### PR DESCRIPTION
**Pull request by albrja**
### Assigned guardians to newborns each time step.
- *Category*: Feature
- *JIRA issue*: [MIC-3595](https://jira.ihme.washington.edu/browse/MIC-3595)
- *Research reference*: https://vivarium-research.readthedocs.io/en/latest/concept_models/vivarium_census_synthdata/concept_model.html#initializing-guardian-s-for-all-simulants

-Updated helper function to choose random guardian if multiple by adding additional argument for groupby call.
-Added function to create mothers household structure
-Assign guardian_1 and guardian_2 to newborns.

### Verification and Testing
Successfully ran simulation and saw updated guardian columns for newborns.
new_births_1
new_births_2
new_births_3

### CODE REVIEW

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 1]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
         # member_ids is a subset of child_households dataframe
+        # groupby_level will be index level to group by (the first level index of member_ids (child_id or mother_id).
         member_ids = (
             member_ids.reset_index()
-            .groupby(["child_id"])["person_id"]
+            .groupby([groupby_level])["person_id"]
             .apply(np.random.choice)
         )
         return member_ids
+
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/components/population.py:615]: Everything above here is just a small refactor and black updates.  Everything below here is newly implemented code.

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 2]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
     @staticmethod
-    def choose_random_guardian(member_ids: pd.DataFrame) -> pd.Series:
+    def choose_random_guardian(member_ids: pd.DataFrame, groupby_level: str) -> pd.Series:
         # member_ids is a subset of child_households dataframe
+        # groupby_level will be index level to group by (the first level index of member_ids (child_id or mother_id).
         member_ids = (
             member_ids.reset_index()
-            .groupby(["child_id"])["person_id"]
+            .groupby([groupby_level])["person_id"]
             .apply(np.random.choice)
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/components/population.py:612]: Not part of this PR, but is `np.random.choice` appropriate or does it need to be a predictable randomness stream like we usually do?

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 3]
```diff
@@ -321,9 +325,10 @@ def initialize_newborns(self, pop_data: SimulantData) -> None:
 
         # typing
         new_births["household_id"] = new_births["household_id"].astype(int)
-        # todo: Update guardian columns MIC-3595
-        new_births["guardian_1"] = data_values.UNKNOWN_GUARDIAN_IDX
-        new_births["guardian_2"] = data_values.UNKNOWN_GUARDIAN_IDX
+        # Assign guardian_1 to birth mother
+        new_births["guardian_1"] = new_births["parent_id"]
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/components/population.py:329]: `parent_id` is definitively the mother?

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 4]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+            .set_index(["household_id", "person_id"])
+        )
+
+        mother_households = mothers.merge(household_info, left_index=True, right_index=True)
+        mother_households.rename(
+            columns={
+                "relation_to_household_head_x": "mother_relation_to_household_head",
+                "relation_to_household_head_y": "member_relation_to_household_head",
+            },
+            inplace=True,
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/components/population.py:717]: nit: you `inplace=True` here but then overwrite in the next `.droplevel` step. Consistency is nice

Comment (rmudambi) [reply src/vivarium_census_prl_synth_pop/components/population.py:717]: +1

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 5]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+            reference_person_ids = self.choose_random_guardian(
+                reference_person_ids, "mother_id"
+            )
+            new_births.loc[
+                new_births["parent_id"].isin(reference_person_ids.index), "guardian_2"
+            ] = new_births["parent_id"].map(reference_person_ids)
+
+        return new_births
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (rmudambi) [reply src/vivarium_census_prl_synth_pop/components/population.py:679]: Pycharm encourages you to make methods static if they can be. This function doesn't need access to the `self` object so it can function as a static method. See here for some insights into the benefits - https://stackoverflow.com/questions/2438473/what-is-the-advantage-of-using-static-methods. In this case, it makes very little difference.

File: src/vivarium_census_prl_synth_pop/components/population.py (hunk 2 above)
Comment (rmudambi) [reply src/vivarium_census_prl_synth_pop/components/population.py:612]: Since we're not running multiple scenarios the common random number (CRN) framework is less valuable. Also, since the point is to generate simulated data, the specifics don't really matter as long as they're internally consistent. 

The most important thing is to be able to harness numpy for performance reasons.

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 6]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+        Returns
+        -------
+        new_births with the additional guardian_2 column updated.
+        """
+        # Setup
+        new_births["guardian_2"] = data_values.UNKNOWN_GUARDIAN_IDX
+        mothers_households = self.get_mothers_household_structure(
+            new_births["parent_id"], households
+        )
+        partners = [
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/population.py:635]: I'd suggest defining all these relationship lists as constants outside the functions, since you're using the same list of relationships for partners in both functions now.

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 7]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+            "Opp-sex spouse",
+            "Opp-sex partner",
+            "Same-sex spouse",
+            "Same-sex partner",
+        ]
+
+        # Index helpers
+        # Mother index groups
+        mother_rp_idx = mothers_households.loc[
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/population.py:644]: rp => ref_person

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 8]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+            mothers_households["member_relation_to_household_head"] == "Reference person"
+        ].index
+
+        # Assign second guardian to random partner of mothers
+        partner_ids = mothers_households.loc[mother_rp_idx.intersection(partners_idx)]
+        if len(partner_ids) > 0:
+            partner_ids = self.choose_random_guardian(partner_ids, "mother_id")
+            new_births.loc[
+                new_births["parent_id"].isin(partner_ids.index), "guardian_2"
+            ] = new_births["parent_id"].map(partner_ids)
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/population.py:664]: Why isn't this just
```
new_births.loc[partner_ids.index, "guardian_2"] = partner_ids
```

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 9]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+        reference_person_ids = mothers_households.loc[
+            mother_partner_idx.intersection(reference_person_idx)
+        ]
+        if len(reference_person_ids) > 0:
+            reference_person_ids = self.choose_random_guardian(
+                reference_person_ids, "mother_id"
+            )
+            new_births.loc[
+                new_births["parent_id"].isin(reference_person_ids.index), "guardian_2"
+            ] = new_births["parent_id"].map(reference_person_ids)
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/population.py:675]: Same question as above

**Review by rmudambi — APPROVED** [review]
Looks good, I just have a few questions that I'd like answers to at stand-up tomorrow

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 10]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+        if len(partner_ids) > 0:
+            partner_ids = self.choose_random_guardian(partner_ids, "mother_id")
+            new_births.loc[
+                new_births["parent_id"].isin(partner_ids.index), "guardian_2"
+            ] = new_births["parent_id"].map(partner_ids)
+
+        reference_person_ids = mothers_households.loc[
+            mother_partner_idx.intersection(reference_person_idx)
+        ]
+        if len(reference_person_ids) > 0:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/components/population.py:669]: Consider making a helper function since these two `if` statements are identical.

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 11]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+            reference_person_ids = self.choose_random_guardian(
+                reference_person_ids, "mother_id"
+            )
+            new_births.loc[
+                new_births["parent_id"].isin(reference_person_ids.index), "guardian_2"
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/components/population.py:674]: So if the parent id is not in the curated list of possible guardians then it just stays with data_values.UNKNOWN_GUARDIAN_IDX? I would have thought it would go to the next family member on some hierarchy of something like that (but I haven't read the docs in depth either)

File: src/vivarium_census_prl_synth_pop/components/population.py (hunk 3 above)
Comment (albrja) [reply src/vivarium_census_prl_synth_pop/components/population.py:329]: Yes.  That how it was defined in earlier versions before the guardian functionality was known to us.

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 12]
```diff
@@ -594,11 +603,119 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+            mothers_households["member_relation_to_household_head"].isin(partners)
+        ].index
+        reference_person_idx = mothers_households.loc[
+            mothers_households["member_relation_to_household_head"] == "Reference person"
+        ].index
+
+        # Assign second guardian to random partner of mothers
+        partner_ids = mothers_households.loc[mother_rp_idx.intersection(partners_idx)]
+        if len(partner_ids) > 0:
+            partner_ids = self.choose_random_guardian(partner_ids, "mother_id")
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/components/population.py:661]: This needs a method name change if it's now used for something other than choosing random guardians

File: src/vivarium_census_prl_synth_pop/components/population.py (hunk 8 above)
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/components/population.py:664]: So if the partner of a a mother does not live in the household then the child gets assigned a guardian of `data_values.UNKNOWN_GUARDIAN_IDX`? It seems like there would be some functionality to have non-partner guardians but I admittedly haven't read through the docs

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 13]
```diff
@@ -577,28 +575,139 @@ def get_household_structure(pop: pd.DataFrame) -> pd.DataFrame:
...
+non_relatives = ["Roommate", "Other nonrelative"]
+children = ["Biological child", "Adopted child", "Foster child", "Stepchild"]
+children_relatives = ["Sibling", "Other relative", "Grandchild", "Child-in-law"]
+parents = ["Parent", "Parent-in-law"]
+partners = [
+    "Opp-sex spouse",
+    "Opp-sex partner",
+    "Same-sex spouse",
+    "Same-sex partner",
+]
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/components/population.py:713]: @beatrixkh Seems likely these will be useful later on for observers or other components so if you think they should live in data values I can move them there.

File: src/vivarium_census_prl_synth_pop/components/population.py (hunk 12 above)
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/components/population.py:642]: Oh I'm confused then. I thought this was choosing a random partner for the female which then may or may not be assigned as guardian of the child.

Comment (albrja) [reply src/vivarium_census_prl_synth_pop/components/population.py:642]: choose_random_guardian is picking a random guardian from a subset of mother_households here (or in the other use cases child_households).  This is why there is always a preceding check for len since you can't groupby on an object of length 0.
